UCP/PROTOV2: support IOV over RDMA_READ in AM/RNDV

jeynmann opened the pull request:
## What
Add IOV support in RTS/RDMA_READ

## Why ?
It might improve the performance of the AM/RNDV protocol, because of currently using b-copy for IOV-type buffer.

## How ?
Pack IOV-type buffer's r-key information in RTS message.
Add remote IOV-info in request and deal IOV-info in GET-ZCOPY protocol.

JeynmannZ committed:
UCP/PROTOV2: support IOV over RDMA_READ in AM/RNDV

JeynmannZ committed:
UCP/PROTOV2: support IOV over RDMA_READ in AM/RNDV

JeynmannZ committed:
UCP/PROTOV2: remove temp logs

JeynmannZ committed:
UCP/PROTOV2: support IOV over RDMA_READ in AM/RNDV

fix gtest:
1. set null rma in send senario;
2. set lane shift 0 when iov size is less than cover max payload;
3. set correct rkey index of next rkey in get-zcopy.

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

fix gtest:
1. fix v1 to v2 senario
2. fix RTS/RDMA_READ(IOV) rkey assert

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

format: adjust format of some files

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

fix gtest:
1. reg mem when md_map is zero in IOV senario
2. fix compile error in RHEL76

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

fix gtest:
1. fix different rkey get same cache

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

Revert "UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV"
This reverts commit a9e1bdcb365dc250773d4a8a9e7354916f62f7c7.

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

fix gtest:
1. fix proto select with rkey(IOV)
2. fix request size test

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RND

fix: add empty iov check

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RND

fix gtest:
1.fix 0-length iov case

leibin2014 commented:
/azp run

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RND

fix gtest:
1. fix timeout if no reg mem when iov is 0. try fallback to contig.
(rc/test_ucp_perf.envelope/31 <rc_v/am_lat_b>)
2. fix rtr init mdmap when buf is iov.

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RND

try fix cuda offload case

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RND

fix gtest: ppln rma_count init

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RND

1.reduce the ucp_request size
2.format

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RN

remove nouse code.
add macro for multirkey reset.

jeynmann commented:
@gleon99 
I think it's ready for review as all g-test are passed. Besides, I've formatted the codes except the old.

ivankochin commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -762,6 +771,243 @@ void ucp_proto_rndv_receive_start(ucp_worker_h worker, ucp_request_t *recv_req,
+        ucp_datatype_iter_cleanup(&req->send.state.dt_iter, UCP_DT_MASK_ALL);
+        goto release_request;
+    }
+    if (req->send.proto_config->proto != &ucp_rndv_get_zcopy_proto) {
```
Wouldn't it better to skip all other protos if `rma_count != 0` instead of this fallback logic. It seems little-bit tricky from my point of view. @yosefe @gleon99 WDYT?

jeynmann replied:
@ivankochin 
Can we do this in stage 3? The current problem is that it cannot get remote `iov`'s count when try to select proto with algorithm V2. Because the proto select parameter does contains all remote address and `rkey`.

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -295,4 +295,78 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+
+    /* 
+     * proto pack format:
+     * | count(32/64)| address(64)| length(32/64)| rkey size(32)| rkey buffer|
```
@ivankochin 
I'm not sure about this. Did we redefine the `size_t` on 32-bit host? Or is it always 64-bit length?

ivankochin replied:
Yes, I understand that `size_t` can be 32 or 64 on different archs, but as I can see count has `uint32_t` type.

jeynmann replied:
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -36,8 +38,10 @@ ucp_proto_rndv_ctrl_get_md_map(const ucp_proto_rndv_ctrl_init_params_t *params,
-
-    if (params->super.super.select_param->dt_class != UCP_DATATYPE_CONTIG) {
+    dt_type      = params->super.super.select_param->dt_class;
+    if (dt_type != UCP_DATATYPE_CONTIG &&
+        !(dt_type == UCP_DATATYPE_IOV &&
+          params->remote_op_id == UCP_OP_ID_RNDV_RECV)) {
         return;
     }
```
done.
I thought this would be adjust again if we add RNDV op in the future.

ivankochin replied:
What do you mean under "RNDV op"?

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -295,4 +295,78 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
     return packed_rkey_size;
```
done.

jeynmann replied:
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -295,4 +295,78 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+    }
+
+    /* 
+     * proto pack format:
```
done

jeynmann replied:
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -295,4 +295,78 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+                                               &dt_iter->mem_info,
+                                               distance_dev_map, dev_distance,
+                                               p);
+        ucs_debug("index=%u rkey_size=%ld address=%p size=%lu",
```
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -295,4 +295,78 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+        *ppacked_rkey_size = packed_rkey_size;
+        p = UCS_PTR_BYTE_OFFSET(p, packed_rkey_size);
+
+        total_size += sizeof(uint64_t) + sizeof(uint32_t) + sizeof(size_t) +
+                      packed_rkey_size;
```
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -295,4 +295,78 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+        total_size += sizeof(uint64_t) + sizeof(uint32_t) + sizeof(size_t) +
+                      packed_rkey_size;
+    }
+    if (packed_count != 0) {
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -762,6 +771,243 @@ void ucp_proto_rndv_receive_start(ucp_worker_h worker, ucp_request_t *recv_req,
+                                                    size_t length,
+                                                    ucp_request_t *req)
+{
+    const void *p              = buffer, *p_end;
```
done.

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -762,6 +771,243 @@ void ucp_proto_rndv_receive_start(ucp_worker_h worker, ucp_request_t *recv_req,
+    size_t remote_size, accumulate_size;
+    ucs_status_t status;
+
+    p_end = UCS_PTR_BYTE_OFFSET(p, length);
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -762,6 +771,243 @@ void ucp_proto_rndv_receive_start(ucp_worker_h worker, ucp_request_t *recv_req,
+    p_end = UCS_PTR_BYTE_OFFSET(p, length);
+    p     = ucp_proto_rndv_unpack_iov_count(p, &iov_count);
+    if (iov_count == 0) {
+        ucs_info("iov count is 0");
+        status = UCS_ERR_NO_MESSAGE;
+        goto finish;
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -762,6 +771,243 @@ void ucp_proto_rndv_receive_start(ucp_worker_h worker, ucp_request_t *recv_req,
+    req->send.rndv.rma_array = ucs_malloc(
+            iov_count * sizeof(*req->send.rndv.rma_array), "proto_rndv");
+    if (req->send.rndv.rma_array == NULL) {
+        ucs_error("failed to allocate remote key");
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -762,6 +771,243 @@ void ucp_proto_rndv_receive_start(ucp_worker_h worker, ucp_request_t *recv_req,
+
+release_request:
+    ucs_mpool_put(req);
+    return;
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -762,6 +771,243 @@ void ucp_proto_rndv_receive_start(ucp_worker_h worker, ucp_request_t *recv_req,
+    if (ucs_unlikely(rts->address == 0 && rkey_length != 0)) {
+        ucp_proto_rndv_receive_iov_start(worker, recv_req, rts, rkey_buffer,
+                                         rkey_length);
+        return;
```
done

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RN

adjust codes according to the review comments

jeynmann commented on src/ucp/rndv/proto_rndv.inl:
```diff
@@ -101,20 +101,47 @@ static UCS_F_ALWAYS_INLINE size_t ucp_proto_rndv_rts_pack(
+    rts->address = 0;
+    rkey_size    = 0;
+pack_done:
     return hdr_len + rkey_size;
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.inl:
```diff
@@ -101,20 +101,47 @@ static UCS_F_ALWAYS_INLINE size_t ucp_proto_rndv_rts_pack(
+    size_t iov_packed_size, iov_count;
+    const ucp_proto_rndv_ctrl_priv_t *rpriv;
+
+    rpriv = req->send.proto_config->priv;
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.inl:
```diff
@@ -153,6 +205,18 @@ ucp_proto_rndv_rkey_destroy(ucp_request_t *req)
 }
 
+static UCS_F_ALWAYS_INLINE void
+ucp_proto_rndv_all_rkey_destroy(ucp_request_t *req)
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.inl:
```diff
@@ -161,6 +225,14 @@ ucp_proto_rndv_recv_complete_with_ats(ucp_request_t *req, uint8_t ats_stage)
 }
 
+static UCS_F_ALWAYS_INLINE void
+ucp_proto_rndv_recv_all_complete_with_ats(ucp_request_t *req, uint8_t ats_stage)
```
done

jeynmann commented on src/ucp/rndv/rndv_get.c:
```diff
@@ -79,13 +79,62 @@ ucp_proto_rndv_get_common_request_init(ucp_request_t *req)
+                                                   lpriv->super.lane);
+        return ucp_rkey_get_tl_rkey(rkey, rkey_index);
+    }
+    return ucp_rkey_get_tl_rkey(req->send.rndv.rkey, lpriv->super.rkey_index);
```
done

jeynmann committed:
Merge branch 'master' into rndv_iov

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -762,6 +771,243 @@ void ucp_proto_rndv_receive_start(ucp_worker_h worker, ucp_request_t *recv_req,
+                                         rkey_length);
+        return;
+    }
+    ucp_proto_rndv_receive_contig_start(worker, recv_req, rts, rkey_buffer,
```
done

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

1. fix merge
2. format code

ivankochin commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -631,15 +637,180 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+{
+    ucp_ep_config_t *ep_config = ucp_ep_config(ep);
+
+    if (length == 0) {
+        goto unpack_nothing;
+    }
+
+    ucs_assert(buffer != NULL);
+
+    if (remote_address != 0) {
+        ucp_proto_rndv_check_rkey_length(remote_address, length, "rts");
+        req->send.rndv.remote_address = remote_address;
+        return ucp_proto_rndv_unpack_contig_rkey(buffer, length, ep, ep_config,
+                                                 req);
+    }
+
+    if (req->send.state.dt_iter.dt_class != UCP_DATATYPE_CONTIG ||
+        remote_size == 0) {
+        goto unpack_nothing;
+    }
+
+    return ucp_proto_rndv_unpack_iov_rkeys(buffer, length, ep, ep_config, req);
+
+unpack_nothing:
+    req->send.rndv.rma_count      = 0;
+    req->send.rndv.remote_address = remote_address;
+    req->send.rndv.rkey           = NULL;
+    return UCS_OK;
```
Would something like that work?

```suggestion
    if (length == 0 || req->send.state.dt_iter.dt_class == UCP_DATATYPE_GENERIC || remote_size == 0) {
        req->send.rndv.rma_count      = 0;
        req->send.rndv.remote_address = remote_address;
        req->send.rndv.rkey           = NULL;
        return UCS_OK;
    }

    ucs_assert(buffer != NULL);

    if (req->send.state.dt_iter.dt_class == UCP_DATATYPE_CONTIG) {
        ucp_proto_rndv_check_rkey_length(remote_address, length, "rts");
        req->send.rndv.remote_address = remote_address;
        return ucp_proto_rndv_unpack_contig_rkey(buffer, length, ep, ep_config,
                                                 req);
    }

    ucs_assert(req->send.state.dt_iter.dt_class == UCP_DATATYPE_IOV);
    return ucp_proto_rndv_unpack_iov_rkeys(buffer, length, ep, ep_config, req);
```

jeynmann replied:
No. Those if blocks are in the receiver side. I think we don't support IOV in RNDV-GET currently.

ivankochin commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -631,15 +637,180 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+                                                 req);
+    }
+
+    if (req->send.state.dt_iter.dt_class != UCP_DATATYPE_CONTIG ||
```
How do we reach `ucp_proto_rndv_unpack_iov_rkeys` if this `if` is `true` for `UCP_DATATYPE_IOV`?

jeynmann replied:
The sender side is support IOV, while the receiver side is not.
If we need IOV support in the receiver side, it will have additional work for the RNDV-GET proto.

ivankochin commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -631,15 +637,180 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_proto_rndv_unpack_common_rkey(
+        const void *buffer, size_t length, uint64_t remote_address,
```
I would prefer to keep consistent with `ucp_proto_rndv_receive_start` names
```suggestion
        const void *rkey_buffer, size_t rkey_length, uint64_t remote_address,
```

jeynmann replied:
done

ivankochin commented on src/ucp/rndv/rndv_get.c:
```diff
@@ -128,6 +179,37 @@ ucp_proto_rndv_get_zcopy_query(const ucp_proto_query_params_t *params,
+    if (req->send.rndv.rma_count == 0) {
+        return max_payload;
+    }
+    iov_index = req->send.rndv.rma_index;
```
What test and with what parameters should I run to reach this point? I tried `test_ucp_am_nbx_rndv.dts` but it always chooses other protos.

jeynmann replied:
If use `UCX_PROTO_ENABLE=y UCX_RNDV_SCHEME=get_zcopy UCX_RNDV_THRESH=0`, it will choose `rndv_get` for iov definitely. And if the message length is zero, it will fail when register memory handler and cause other unexpected behaviors.

jeynmann replied:
If use `UCX_PROTO_ENABLE=y UCX_RNDV_SCHEME=get_zcopy UCX_RNDV_THRESH=0`, it will choose `rndv_get` for iov definitely.

ivankochin replied:
Is it possible to create very simple and straightforward test-case expressed just in several lines that covers this new flow?

E.g. add test case to `test_ucp_am_nbx_rndv` with implied `UCX_PROTO_ENABLE=y UCX_RNDV_SCHEME=get_zcopy UCX_RNDV_THRESH=0` environment.

The diff in this patch is already around 500 lines so I would like to include testing in that patch only if it is very straightforward and simple change.

jeynmann replied:
done

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

adjust codes according to the comments

jeynmann commented on src/ucp/core/ucp_request.h:
```diff
@@ -234,11 +234,27 @@ struct ucp_request {
+                        };
+                    };
+                    /* Key counts for remote buffer operation */
+                    size_t rma_count;
```
done

ivankochin replied:
I would also propose to rename `remote_address` and `remote_size` to `address` and `size` accordingly. WDYTA?

jeynmann replied:
done

ivankochin replied:
Sorry, misunderstanding happened.

I meant to change `remote_address` and `remote_size` to `address` and `size` without `r` prefix since they are already inside `rdata` member that already contains this `r` part.

jeynmann replied:
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -302,4 +299,79 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+{
+    const ucp_datatype_iter_t *dt_iter = &req->send.state.dt_iter;
+    ucp_mem_h *memh = dt_iter->type.iov.memh;
+    void *p = rkey_buffer;
```
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -302,4 +299,79 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
 }
 
+static UCS_F_ALWAYS_INLINE size_t
+ucp_proto_request_pack_rkey_iov(ucp_request_t *req, ucp_md_map_t md_map,
```
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -275,17 +275,14 @@ ucp_proto_request_send_op(ucp_ep_h ep, ucp_proto_select_t *proto_select,
-                            uint64_t distance_dev_map,
-                            const ucs_sys_dev_distance_t *dev_distance,
-                            void *rkey_buffer)
+ucp_proto_request_pack_rkey_contig(ucp_request_t *req, ucp_md_map_t md_map,
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -23,6 +24,7 @@ ucp_proto_rndv_ctrl_get_md_map(const ucp_proto_rndv_ctrl_init_params_t *params,
 {
     ucp_context_h context                    = params->super.super.worker->context;
     const ucp_ep_config_key_t *ep_config_key = params->super.super.ep_config_key;
+    uint8_t dt_type;
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -631,15 +637,180 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+        * rkey_ptr_lane. Examples are: sysv and posix. Such keys, if packed,
+        * are unpacked only once and cached in the peer_mem hash on ep. It is
+        * done by the specific protocols (if selected) which use them.
+        */
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -695,28 +858,16 @@ ucp_proto_rndv_send_reply(ucp_worker_h worker, ucp_request_t *req,
                                   const void *rkey_buffer, size_t rkey_length)
 {
+    uint64_t remote_address = rts->address;
+    size_t remote_size      = rts->size;
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -649,27 +820,18 @@ ucp_proto_rndv_send_reply(ucp_worker_h worker, ucp_request_t *req,
-        }
-
+    if (req->send.rndv.rkey == NULL) {
+        proto_select   = &ucp_ep_config(ep)->proto_select;
```
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -302,4 +299,79 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+     */
+    ppacked_count = (uint32_t*)p;
+    p             = UCS_PTR_BYTE_OFFSET(p, sizeof(uint32_t));
+    iov_count     = ucp_datatype_iter_iov_count(dt_iter);
```
done

jeynmann commented on src/ucp/dt/datatype_iter.inl:
```diff
@@ -47,6 +60,10 @@ ucp_datatype_iov_iter_init(ucp_context_h context, void *buffer, size_t count,
 {
     const ucp_dt_iov_t *iov = (const ucp_dt_iov_t*)buffer;
+    if (length == 0) {
+        ucp_datatype_iter_init_null(dt_iter, length, sg_count);
```
If the message length is zero, it will fail when register memory handler (debug version) and cause the `gtest` case failed.

jeynmann committed:
Merge branch 'master' into rndv_iov

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

rename rdata's member

jeynmann commented on src/ucp/rndv/proto_rndv.inl:
```diff
@@ -99,20 +99,42 @@ static UCS_F_ALWAYS_INLINE size_t ucp_proto_rndv_rts_pack(
+    iov_packed_size = sizeof(uint64_t) + sizeof(size_t) + sizeof(uint32_t) +
+                      rpriv->packed_rkey_size;
+    iov_count       = ucp_datatype_iter_iov_count(&req->send.state.dt_iter);
+    return sizeof(ucp_rndv_rts_hdr_t) + sizeof(uint32_t) +
```
done

jeynmann committed:
Merge branch 'master' into rndv_iov

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

fix gtest fail 'rc_verbs_ep.c:253  UCX  ERROR Invalid get_zcopy length: 56 (expected: >= 65)'

ivankochin commented on src/ucp/rndv/rndv_get.c:
```diff
@@ -140,16 +220,16 @@ ucp_proto_rndv_get_zcopy_send_func(ucp_request_t *req,
     ucs_assert(iov.count == 1);
     ucp_proto_common_zcopy_adjust_min_frag(req, rpriv->mpriv.min_frag,
-                                           iov.length, &iov, 1, &offset);
+                                            iov.length, &iov, 1, &offset);
```
Seems like broken indentation

jeynmann replied:
done

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

proto select with all rkeys

JeynmannZ committed:
UCP/PROTO2: support IOV over RDMA_READ in AM/RNDV

adjust code

jeynmann committed:
Merge branch 'openucx:master' into rndv_iov

jeynmann commented on src/ucp/rndv/proto_rndv.inl:
```diff
@@ -145,7 +176,32 @@ static size_t UCS_F_ALWAYS_INLINE ucp_proto_rndv_ack_progress(
+    req->send.rndv.rdata       = NULL;
+    req->send.rndv.rdata_count = 0;
+#if UCS_ENABLE_ASSERT
+    req->send.rndv.rdata_idx   = 0;
```
Done.
`rdate_idx` is stored in the same block with `rkey`.
`rdate_idx` is not necessary to be reset.
Anyway, I removed this for resetting the value doesn't cost too much..

see:
``` c
static UCS_F_ALWAYS_INLINE void
ucp_proto_rndv_contig_rkey_destroy(ucp_request_t *req)
{
    ucs_assert(req->send.rndv.rkey != NULL);
    ucp_rkey_destroy(req->send.rndv.rkey);
#if UCS_ENABLE_ASSERT
    req->send.rndv.rkey = NULL;
#endif
}
```

JeynmannZ committed:
UCP/PROTOV2: support IOV over RDMA_READ in AM/RNDV

rename raddres risze of request

JeynmannZ committed:
UCP/PROTOV2: support IOV over RDMA_READ in AM/RNDV

1.fix adjust payload
2.add gtest case

JeynmannZ committed:
UCP/PROTOV2: support IOV over RDMA_READ in AM/RNDV

1.fix offset
2.format

JeynmannZ committed:
UCP/PROTOV2: support IOV over RDMA_READ in AM/RNDV

1.format

jeynmann commented on src/ucp/rndv/proto_rndv.inl:
```diff
@@ -90,30 +90,61 @@ ucp_proto_rndv_ats_handler(void *arg, void *data, size_t length, unsigned flags)
         ucp_request_t *req, ucp_rndv_rts_hdr_t *rts, size_t hdr_len)
 {
-    void *rkey_buffer = UCS_PTR_BYTE_OFFSET(rts, hdr_len);
+    void *rkey_buffer   = UCS_PTR_BYTE_OFFSET(rts, hdr_len);
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.inl:
```diff
@@ -90,30 +90,61 @@ ucp_proto_rndv_ats_handler(void *arg, void *data, size_t length, unsigned flags)
+    iov_count       = ucp_datatype_iter_iov_count(&req->send.state.dt_iter);
+    max_rts_size    = sizeof(ucp_rndv_rts_hdr_t) + sizeof(size_t) +
+                      iov_packed_size * iov_count;
+    if (max_rts_size > max_bcopy_size) {
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -631,14 +636,178 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+            goto destroy_rkey;
+        }
+        accumulate_size                                    += rsize;
+        req->send.rndv.rdata[iov_index].accumulate_size = accumulate_size;
```
done

jeynmann commented on src/ucp/rndv/proto_rndv.inl:
```diff
@@ -90,30 +90,61 @@ ucp_proto_rndv_ats_handler(void *arg, void *data, size_t length, unsigned flags)
-        rts->address = 0;
-        rkey_size    = 0;
-    } else {
+    if (ucp_proto_rndv_max_rts_size(req) == sizeof(ucp_rndv_rts_hdr_t) ||
```
We need to check `max_rts_size` again to decide whether to pack `rkeys` or not.

brminich commented on src/ucp/proto/proto_common.inl:
```diff
@@ -371,4 +368,79 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+     *               | address(64)| length(32/64)| rkey size(32/64)| rkey buffer|
+     */
+    ppacked_count = (size_t*)p;
+    p             = UCS_PTR_BYTE_OFFSET(p, sizeof(size_t));
```
pls use uint32_t or uin64_t explicitly when packing/unpacking the data which is intended to be transferred between different peers

jeynmann replied:
done

brminich commented on src/ucp/proto/proto_common.inl:
```diff
@@ -371,4 +368,79 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+    ssize_t packed_rkey_size;
+    size_t total_size;
+
+    ucs_assert(dt_iter->dt_class == UCP_DATATYPE_IOV);
```
pls use assertv and print the real class

jeynmann replied:
done

brminich commented on src/ucp/proto/proto_common.inl:
```diff
@@ -371,4 +368,79 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+    total_size    = 0;
+    for (iov_index = 0; iov_index != iov_count; ++iov_index) {
+        if (dt_iter->type.iov.iov[iov_index].length == 0) {
+            ucs_info("iov[%lu] block size is 0", iov_index);
```
pls use `ucs_data` as it is supposed to be printed for every msg
`ucs_info` is for important notifications printed quite rarely

jeynmann replied:
done

brminich commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -36,8 +38,11 @@ ucp_proto_rndv_ctrl_get_md_map(const ucp_proto_rndv_ctrl_init_params_t *params,
+    if (dt_type == UCP_DATATYPE_GENERIC) {
+        return;
+    }
+    if (dt_type == UCP_DATATYPE_IOV &&
+        params->remote_op_id != UCP_OP_ID_RNDV_RECV) {
```
```suggestion
    if ((dt_type == UCP_DATATYPE_IOV) &&
        (params->remote_op_id != UCP_OP_ID_RNDV_RECV)) {
```

jeynmann replied:
done

brminich commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -631,14 +636,179 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+    req->send.rndv.rdata = ucs_malloc(iov_count * sizeof(*req->send.rndv.rdata),
+                                      "proto_rndv");
+    if (req->send.rndv.rdata == NULL) {
+        ucs_error("failed to allocate rma array");
```
`array for iov rkeys`?

jeynmann replied:
Yes, but not only rkeys, but also iov length and address.

brminich commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -631,14 +636,179 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+
+    ucs_assert(rkey_buffer != NULL);
+
+    if (remote_address != 0 || remote_size == 0) {
```
```suggestion
    if ((remote_address != 0) || (remote_size == 0)) {
```

jeynmann replied:
done

brminich commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -631,14 +636,179 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+                                                 ep_config, req);
+    }
+
+    if (req->send.state.dt_iter.dt_class != UCP_DATATYPE_CONTIG) {
```
why?

brminich reviewed (changes requested):
imo, the general approach is not correct here. We can't assume that any peer will be able to handle new RTS format with IOVs. We need to introduce new version of UCP protocols and exchange this info during wireup (in worker address). If both peers can handle RNDV with IOVs, then we can use it, otherwise fallback to the existing AM-based scheme

JeynmannZ committed:
UCP/PROTOV2: support IOV over RDMA_READ in AM/RNDV

1. use uint64_t instead of size_t
2. use assertv

brminich commented:
> 

wire compat is not an issue, since this is fo proto v2 only

jeynmann commented on src/ucp/rndv/proto_rndv.c:
```diff
@@ -23,6 +24,7 @@ ucp_proto_rndv_ctrl_get_md_map(const ucp_proto_rndv_ctrl_init_params_t *params,
 {
     ucp_context_h context                    = params->super.super.worker->context;
     const ucp_ep_config_key_t *ep_config_key = params->super.super.ep_config_key;
+    uint8_t dt_type = params->super.super.select_param->dt_class;
```
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -371,4 +368,79 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+                  (void*)dt_iter->type.iov.iov[iov_index].buffer,
+                  dt_iter->type.iov.iov[iov_index].length);
+
+        if (packed_rkey_size < 0) {
```
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -371,4 +368,79 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+        *(uint64_t*)p = (uintptr_t)dt_iter->type.iov.iov[iov_index].buffer;
+        p             = UCS_PTR_BYTE_OFFSET(p, sizeof(uint64_t));
+        *(size_t*)p   = dt_iter->type.iov.iov[iov_index].length;
+        p             = UCS_PTR_BYTE_OFFSET(p, sizeof(size_t));
```
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -371,4 +368,79 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+        p             = UCS_PTR_BYTE_OFFSET(p, sizeof(size_t));
+
+        ppacked_rkey_size = (size_t*)p;
+        p                 = UCS_PTR_BYTE_OFFSET(p, sizeof(size_t));
```
done

jeynmann commented on src/ucp/proto/proto_common.inl:
```diff
@@ -371,4 +368,79 @@ ucp_proto_request_pack_rkey(ucp_request_t *req, ucp_md_map_t md_map,
+    size_t total_size;
+
+    ucs_assert(dt_iter->dt_class == UCP_DATATYPE_IOV);
+    if (md_map == 0 || dt_iter->length == 0) {
```
done

gleon99 commented:
Hi @jeynmann - we're cleaning up the UCX PR backlog and plan to close this PR in 7 days. Please let us know if you'd like to keep it open.
